File: mgmt-application/src/main/java/io/quartic/mgmt/CreateStaticDatasetRequest.java
package io.quartic.mgmt;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import io.quartic.catalogue.api.DatasetMetadata;
import io.quartic.common.SweetStyle;
import org.immutables.value.Value;

@SweetStyle
@Value.Immutable
@JsonSerialize(as = CreateStaticDatasetRequestImpl.class)
@JsonDeserialize(as = CreateStaticDatasetRequestImpl.class)
public interface CreateStaticDatasetRequest extends CreateDatasetRequest {
    DatasetMetadata metadata();
    String fileName();
    FileType fileType();

    @Override
    default <T> T accept(Visitor<T> visitor) {
        return visitor.visit(this);
    }
}

File: mgmt-application/src/main/kotlin/io/quartic/mgmt/CreateStaticDatasetRequest.kt
package io.quartic.mgmt

import io.quartic.catalogue.api.model.DatasetMetadata
import io.quartic.catalogue.api.model.MimeType

data class CreateStaticDatasetRequest(
        val metadata: DatasetMetadata,
        val fileName: String,
        val fileType: FileType
) : CreateDatasetRequest {
    fun mimeType() = when (fileType) {
        FileType.CSV -> MimeType.CSV
        FileType.GEOJSON -> MimeType.GEOJSON
        FileType.RAW -> MimeType.RAW
    }
}
